### ISSUE TITLE
HttpClient.GetAsync() causes uncatchable CryptographicException on Linux in OpenSslX509ChainProcessor.MapVerifyErrorToChainStatus

### ISSUE THREAD

**Issue by Qowy** [dotnet/corefx#21793]
System: Ubuntu 16.04 and Ubuntu 16.10
Dotnet release: dotnet-dev-1.0.4

apt list --installed | grep dotnet output:
```
dotnet-dev-1.0.4/yakkety,now 1.0.4-1 amd64 [installed]
dotnet-host/yakkety,now 2.0.0-preview2-25407-01-1 amd64 [installed,automatic]
dotnet-hostfxr-1.1.0/yakkety,now 1.1.0-1 amd64 [installed,automatic]
dotnet-sharedframework-microsoft.netcore.app-1.1.2/yakkety,now 1.1.2-1 amd64 [installed,automatic]
```

apt list --installed | grep openssl output:
```
libgnutls-openssl27/yakkety-updates,yakkety-security,now 3.5.3-5ubuntu1.2 amd64 [installed]
openssl/yakkety-updates,now 1.0.2g-1ubuntu9.3 amd64 [installed]
```

Repro Code:
```
using System;
using System.Net.Http;
using System.Threading.Tasks;

namespace Repro
{
    class Program
    {
        static void Main(string[] args)
        {
            Do().GetAwaiter().GetResult();
        }

        static async Task Do()
        {
            var client = new HttpClient();
            try
            {
                var response = await client.GetAsync((new Uri("https://obr03.ru/")));
                Console.WriteLine("Ok");
            }
            catch(Exception ex)
            {
                Console.WriteLine("Catched");
            }
        }
    }
}
```

Result Windows:
`Catched`

Result Linux:
```
Unhandled Exception: System.Security.Cryptography.CryptographicException: Error occurred during a cryptographic operation.
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.MapVerifyErrorToChainStatus(X509VerifyStatusCode code)
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.AddElementStatus(X509VerifyStatusCode errorCode, List`1 elementStatus, List`1 overallStatus)
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.AddElementStatus(List`1 errorCodes, List`1 elementStatus, List`1 overallStatus)
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.BuildChain(X509Certificate2 leaf, HashSet`1 candidates, HashSet`1 downloaded, HashSet`1 systemTrusted, OidCollection applicationPolicy, OidCollection certificatePolicy, X509RevocationMode revocationMode, X509RevocationFlag revocationFlag, DateTime verificationTime, TimeSpan& remainingDownloadTime)
   at Internal.Cryptography.Pal.ChainPal.BuildChain(Boolean useMachineContext, ICertificatePal cert, X509Certificate2Collection extraStore, OidCollection applicationPolicy, OidCollection certificatePolicy, X509RevocationMode revocationMode, X509RevocationFlag revocationFlag, DateTime verificationTime, TimeSpan timeout)
   at System.Security.Cryptography.X509Certificates.X509Chain.Build(X509Certificate2 certificate)
   at System.Net.Security.CertificateValidation.BuildChainAndVerifyProperties(X509Chain chain,  509Certificate2 remoteCertificate, Boolean checkCertName, String hostName)
   at System.Net.Http.CurlHandler.SslProvider.VerifyCertChain(IntPtr storeCtxPtr, IntPtr curlPtr)
   at Interop.Http.MultiPerform(SafeCurlMultiHandle multiHandle)
   at System.Net.Http.CurlHandler.MultiAgent.WorkerBodyLoop()
Aborted (core dumped)
```

**Comment by davidsh** [comment]
The TLS/SSLS certificate is invalid. 

See: https://www.ssllabs.com/ssltest/analyze.html?d=obr03.ru

![image](https://user-images.githubusercontent.com/7684835/27765980-0cfc238e-5e76-11e7-8451-5a8a68b16116.png)


But, regardless, HttpClient is supposed to only throw an HttpRequestException at the top-level. The CryptographicException should be an inner exception.

So, the bug here is not that an exception was thrown, but rather the wrong type of exception.

cc: @stephentoub @bartonjs

**Comment by Qowy** [comment]
Yes I realize this website is malformed but working on something similar to a crawler I cannot choose the websites I work with.

However isn't the even bigger issue that this terminates the execution of the program (see "Aborted (core dumped)" despite the try catch block.

**Comment by davidsh** [comment]
>However isn't the even bigger issue that this terminates the execution of the program (see "Aborted (core dumped)" despite the try catch block.

Yes, that is part of the bug that will be investigated.

**Comment by Qowy** [comment]
Ok thank you for looking into this, any quick and dirty workarounds that come to mind? :D (For now I simply blacklisted this domain, but there may be others)

**Comment by davidsh** [comment]
You can try changing this:

```c#
var client = new HttpClient();
```

to this:

```c#
var handler = new HttpClientHandler();
handler.ServerCertificateCustomValidationCallback  = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator ;
var client = new HttpClient(handler);
```

This will result in ignoring all TLS/SSL certificate errors.

**Comment by davidsh** [comment]
Additionally, this particular website (after ignoring the name mismatch on the TLS/SSL certificate has other problems.  It used a nonstandard cipher suite:

![image](https://user-images.githubusercontent.com/7684835/27766027-a18c3308-5e77-11e7-9ef8-5f59ec1245aa.png)

**Comment by Qowy** [comment]
Yes like I said there are a lot of strangely configured websites out there but that should not crash my code especially with proper exception handling :)
Your suggested workaround sadly seems to be netstandard 2.0 and this project is still on 1.6

Let's hope that this website configuration is rare enough to not cause too much further problems.

**Comment by davidsh** [comment]
The `DangerousAcceptAnyServerCertificateValidator` is part of netstandard 2.0.

However, you can define it yourself like this:

```c#
public static Func<HttpRequestMessage, X509Certificate2, X509Chain, SslPolicyErrors, bool> DangerousAcceptAnyServerCertificateValidator { get; } = delegate { return true; };
    }
}

handler.ServerCertificateCustomValidationCallback = DangerousAcceptAnyServerCertificateValidator;
```

The `ServerCertificateCustomValidationCallback` property should be part of HttpClientHandler in netstandard 1.6.

**Comment by Qowy** [comment]
Ok sadly (or interestingly) this changes nothing.

```
static async Task Do()
        {
            var handler = new HttpClientHandler();
            handler.ServerCertificateCustomValidationCallback = (message, certificate2, arg3, arg4) => true;
            var client = new HttpClient(handler);
            try
            {
                var response = await client.GetAsync((new Uri("https://obr03.ru/")));
                Console.WriteLine("Ok");
            }
            catch(Exception ex)
            {
                Console.WriteLine("Catched");
            }
        }
```

This still leads to a Catched exception on Windows and the error on Linux

**Comment by davidsh** [comment]
>This still leads to a Catched exception on Windows and the error on Linux

Thank you for the extra repro instructions.

I will say that on Windows, there are still some TLS/SSL handshakes errors that can not be overriden despite the "ignore TLS/SSL certificate errors".  Low-level problems such as "unknown cipher suites" as is happening with website cannot be ignored.  At least on Windows, the exception is bubbling up and it is able to be caught in the "catch(Exception ex)" clause.

It looks like on Linux, there is some other bug that is causing a core dump which will need to be investigated.

Overall, though, it is pretty rare for servers to offer an cipher suite that is "unknown" like that.

**Comment by Qowy** [comment]
Ok, again thank you for your time and happy bug hunting. 

Yes it seems to be pretty rare ( so far I had this error once out of about 1m domains), who knows this website may even be intentionally malicious.

Nevertheless this is potentially a serious error. Just think about an asp.net core website with image cross loading or functionality where a user provided url is even just requested once (to check if it is online or whatever).

**Comment by Qowy** [comment]
Ok just to check what I feared this actually crashes the whole kestrel process if used in an ASP.NET core project. Unless I am missing something this could be quite the DoS attack if such a project requests anything from a user supplied URL.

**Comment by stephentoub** [comment]
Have you tried .NET Core 1.1 or 2.0? I'm not at my machine to try, but looking at the code, there's a catch block that should be eating this, so I suspect some kind of exception handling issue in 1.0.
cc: @janvorli

**Comment by Qowy** [comment]
As you can see by the installed packages this with .NET Core 1.1.2

And yes the catch block not doing anything is exactly the issue.
Looking at the Exeption message the stack trace starts with
`System.Net.Http.CurlHandler.MultiAgent.WorkerBodyLoop()`

Which is maybe started by the dotnet cli? This error does not bubble up through my code at all and instead kills the whole dotnet process (as I tested not even ASP.NET cores kestrel server is save from this)

**Comment by stephentoub** [comment]
I was referring to this catch block:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs#L316
which should be catching any exception that emerges from WorkerBodyLoop.

**Comment by stephentoub** [comment]
The stack trace is odd. I think what's happening is an exception is crossing from managed to native code, which isn't supported, and that's wreaking havoc. There should be a catch block in the verify callback that eats the error and returns failure in a supported manner.

**Comment by leecow** [comment]
2.0 pr https://github.com/dotnet/corefx/pull/21888

**Comment by Qowy** [comment]
I don't know id this is the right place to ask, but will 1.0 LTS be replaced by 2.0? Otherwise because of the possible denial of service attack nature of this error on ASP.NET core websites shouldn't this also be backported for the LTS branch?

**Comment by stephentoub** [comment]
> because of the possible denial of service attack nature

I wouldn't characterize this as a DoS avenue.  If a server is accepting and connecting to arbitrary, unvetted urls to fetch, such a server is already vulnerable to all sorts of issues.  And if some site the server is knowingly connecting to is somehow compromised with a maliciously-installed bad certificate, the server can simply stop connecting to it until the target site is fixed (it wouldn't be able to use it for its intended purpose anyway).

**Comment by Qowy** [comment]
I was thinking along the lines of websites that offer analysis of other websites, like ssl configuration etc. But yes this use case is properly rare enough, especially for .NET core that a fix in the 2.0 release may be suficcient
